include/picongpu/fields/background/templates/twtstight/BField.tpp

```diff
@@ -41,273 +36,104 @@ namespace picongpu
-            {
-                pmacc::math::Vector<floatD_X, detail::numComponents> zeroShifts;
-                for(uint32_t component = 0; component < detail::numComponents; ++component)
-                    zeroShifts[component] = floatD_X::create(0.0);
-                return getValue(cellIdx, zeroShifts, currentStep);
-            }
+                auto const [tanPhi, cotPhi, sinPhi_2, sinPolAngle, cosPolAngle]
+                    = defineTrigonometryShortcuts(absPhi, sinPhi);
+                // clang-format off
+                auto const [I, x2, tauG2, psi0, w02, beta02, nu, xi, rhom, Xm, besselI0const, besselJ0const, besselJ1const, zeroOrder]
+                    = defineCommonHelperVariables(absPhi, sinPhi, cosPhi, beta0, tanAlpha, cspeed, lambda0, omega0, tauG, w0, k,
+                                                  x, y, z, t, cotPhi, sinPhi_2);
+                // clang-format on
 
-            HDINLINE
-            float3_X BField::getValue(
-                floatD_X const& cellIdx,
-                pmacc::math::Vector<floatD_X, detail::numComponents> const& extraShifts,
-                float_X const currentStep) const
-            {
-                float_64 const time_SI = float_64(currentStep) * dt - tdelay;
+                /* Calculating shortcuts for speeding up field calculation */
+                float_T const cosPhi_2 = cosPhi * cosPhi;
+                float_T const sin2Phi = math::sin(float_T(2.0) * absPhi);
+                complex_T const rhom2 = rhom * rhom;
 
-                pmacc::math::Vector<floatD_64, detail::numComponents> const bFieldPositions_SI
-                    = detail::getFieldPositions_SI(cellIdx, halfSimSize, extraShifts, unit_length, focus_y_SI, phi);
-                /* Single TWTS-Pulse */
-                return getTWTSBfield_Normalized<simDim>(bFieldPositions_SI, time_SI);
-            }
+                complex_T const result
+                    = (complex_T(0, -0.25) * math::exp(I * (omega0 * t - k * y * cosPhi)) * zeroOrder
+                       * (k * rhom * besselJ0const
+                              * (cosPolAngle * (-rhom2 + x2 + x * cosPhi * Xm) * sinPhi_2
+                                 - sinPolAngle
+                                     * (rhom2 + rhom2 * cosPhi_2 - x2 * sinPhi_2 + x * cosPhi * sinPhi_2 * Xm))
+                          + besselJ1const
+                              * (cosPolAngle * sinPhi
+                                     * (rhom2 - float_T(2.0) * x2 + I * Xm * rhom2 * (k * sinPhi)
+                                        + x * cosPhi * (float_T(-2.0) * Xm - I * rhom2 * (k * sinPhi)))
+                                 + sinPolAngle
+                                     * ((rhom2 - float_T(2.0) * x2) * sinPhi
+                                        + I * rhom2 * (-Xm + x * cosPhi) * (k * sinPhi_2) + x * sin2Phi * Xm)))
+                       * psi0)
+                    / (cspeed * besselI0const * rhom * rhom2);
```

Comment:
I am not entirely convinced, I am still thinking about this.

Reply:
Maybe it is just my taste. I'd argue that it already feels organic to read. Such function extraction would not be about code reuse, but just to get something out of the view.

Reply:
It would be about separation of concerns: The outer function `calcTWTSField`'s concern is providing the interface with convenient arguments by translating them into the variables used by the formula. The new function's concern would be to compute the value from precisely the variables it is formulated with. By separating these concerns the class takes full control over what interface it provides without mangling interface decisions with variable naming conventions. The final function could be as short as:
```C++
auto calcTWTSFieldX(auto pos, ...) {
    auto variables = translateVariables(pos, ...);
    return applyFormula(variables);
}
```
In this formulation, I can easily change the interface of `calcTWTSFieldX` by adjusting `translateVariables` without ever touching something as brittle as 14 lines of difficult math. To me, the latter feels like something you'd really never want to touch again if you got it right once.